File: order-service/src/main/java/com/kidfavor/orderservice/coupon/Coupon.java
package com.kidfavor.orderservice.coupon;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "coupons")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Coupon {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, length = 50)
    private String code;

    @Column(nullable = false)
    private boolean active;

    @Column(nullable = false, precision = 19, scale = 2)
    private BigDecimal discountValue; // either fixed amount or percent

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private DiscountType discountType;

    @Column
    private LocalDateTime expiresAt;

    @Column
    private Integer maxRedemptions;

    @Column(nullable = false)
    @Builder.Default
    private Integer timesRedeemed = 0;

    @Version
    private Long version;

    public enum DiscountType {
        PERCENT, FIXED
    }

    public boolean isExpired() {
        return expiresAt != null && LocalDateTime.now().isAfter(expiresAt);
    }

    public boolean isUsable() {
        if (!active) return false;
        if (isExpired()) return false;
        if (maxRedemptions != null && timesRedeemed >= maxRedemptions) return false;
        return true;
    }
}

File: mobile-app/lib/models/coupon.dart
class Coupon {
  final int id;
  final String code;
  final bool active;
  final double discountValue;
  final String discountType; // PERCENT or FIXED
  final DateTime? expiresAt;
  final int? maxRedemptions;
  final int timesRedeemed;

  Coupon({
    required this.id,
    required this.code,
    required this.active,
    required this.discountValue,
    required this.discountType,
    this.expiresAt,
    this.maxRedemptions,
    required this.timesRedeemed,
  });

  factory Coupon.fromJson(Map<String, dynamic> json) {
    return Coupon(
      id: json['id'] as int? ?? 0,
      code: json['code'] as String? ?? '',
      active: json['active'] as bool? ?? false,
      discountValue: (json['discountValue'] as num?)?.toDouble() ?? 0.0,
      discountType: json['discountType'] as String? ?? 'PERCENT',
      expiresAt: json['expiresAt'] != null
          ? DateTime.tryParse(json['expiresAt'].toString())
          : null,
      maxRedemptions: json['maxRedemptions'] as int?,
      timesRedeemed: json['timesRedeemed'] as int? ?? 0,
    );
  }

  bool get isExpired {
    if (expiresAt == null) return false;
    return DateTime.now().isAfter(expiresAt!);
  }

  bool get isUsable {
    if (!active) return false;
    if (isExpired) return false;
    if (maxRedemptions != null && timesRedeemed >= maxRedemptions!) {
      return false;
    }
    return true;
  }

  // Calculate discount amount based on total
  double calculateDiscount(double total) {
    if (!isUsable) return 0;
    if (discountType == 'PERCENT') {
      return total * (discountValue / 100);
    } else {
      return discountValue;
    }
  }
}
